src/enumeration.rs

```diff
@@ -525,18 +525,28 @@ pub enum UsbController {
     VHCI,
 }
 
-impl UsbController {
+impl UsbControllerType {
     #[allow(dead_code)] // not used on all platforms
     pub(crate) fn from_str(s: &str) -> Option<Self> {
         match s
             .find("HCI")
             .filter(|i| *i > 0)
             .and_then(|i| s.bytes().nth(i - 1))
         {
-            Some(b'x') | Some(b'X') => Some(UsbController::XHCI),
-            Some(b'e') | Some(b'E') => Some(UsbController::EHCI),
-            Some(b'o') | Some(b'O') => Some(UsbController::OHCI),
-            Some(b'v') | Some(b'V') => Some(UsbController::VHCI),
+            Some(b'x') | Some(b'X') => Some(UsbControllerType::XHCI),
+            Some(b'e') | Some(b'E') => Some(UsbControllerType::EHCI),
+            Some(b'o') | Some(b'O') => Some(UsbControllerType::OHCI),
+            Some(b'v') | Some(b'V') => Some(UsbControllerType::VHCI),
+            _ => None,
+        }
+    }
+
+    #[allow(dead_code)] // not used on all platforms
+    pub(crate) fn from_bcd(bcd: u16) -> Option<Self> {
+        match bcd & 0xF00 {
+            0x300 => Some(UsbControllerType::XHCI),
+            0x200 => Some(UsbControllerType::EHCI),
+            0x100 => Some(UsbControllerType::OHCI),
```

Comment:
I suspect the actual value here would be `0x110`. Though good luck finding a system to test it at this point.

Reply:
That's why I applied the `0xF00` mask. Though still not happy that there might be a 4.0 hub or something and a range would be dubious.

Reply:
Ah! I hadn't spotted the mask.